File: arduino/controllerdata.c
#include "controllerdata.h"

enum packet_t { PACKET_A, PACKET_B };

/* Actual network packet */
struct packet {
	uint8_t 	command;
	uint8_t 	static_var;
	uint8_t 	throttle;
	uint8_t 	brake_pedal;
	uint8_t 	brake_switch;
	uint8_t 	foot_switch;
	uint8_t		forward_switch;
	bool 		reverse;
	bool 		hall_a;
	bool 		hall_b;
	bool 		hall_c;
	uint8_t 	battery_voltage;
	uint8_t 	motor_temp;
	uint8_t 	controller_temp;
	bool 		setting_dir;
	bool 		actual_dir;
	uint8_t 	unknown_2;
	uint8_t 	unknown_2;
	uint8_t 	crc;
};

uint8_t 		_calculate_CRC(struct packet *pkt);
bool 		    validate_checksum(struct packet *pkt);
packet_t 		determine_packet_type(struct packet *pkt);


#endif

File: klsreader/protocol/controllerdata.py
from struct import *

class ControllerData():

    def __init__(self, *packets):
        # packet must be 19 bytes
        for packet in packets:

            if self.validateChecksum(packet):
                self.determinePacketType(packet)

    def _calculateCRC(self, data):
        checksum = 0 # checksum is int type
        for ch in data[:18]:
            checksum += ord(ch)
        return checksum % 256 # ensure that the checksum does not exceed 1 byte

    def validateChecksum(self, data):
        # compare received checksum with calculated checksum
        if self._calculateCRC(data) == unpack('B', data[-1])[0]:
            return True
        else:
            # validation will fail if checksum is invalid
            # or if packet is not 19 bytes in length
            return False

    def determinePacketType(self, data):
        # use first byte of packet to determine type
        if data[0] == '\x3A':
            self.type = "A"
            self._unpackPacketA(data)
        elif data[0] == '\x3B':
            self.type = "B"
            self._unpackPacketB(data)
        else:
            # set instance type to "invalid" if packet is of an unknown type
            self.type = "invalid"

    def _unpackPacketA(self, data):
        (self.throttle, self.brakePedal, self.brakeSwitch, self.footSwitch,
        self.forwardSwitch, self.reverse, self.hallA, self.hallB, self.hallC,
        self.batteryVoltage, self.motorTemp, self.controllerTemp, self.settingDir,
        self.actualDir) = unpack('!2x5B4?3B2?3x', data)

    def _unpackPacketB(self, data):
        (self.rpm, self.phaseCurrent) = unpack('!4x2H11x', data)
